lib/libzpool/uzfs_io.c

```diff
@@ -60,6 +64,31 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 
 	rl = zfs_range_lock(&zv->zv_range_lock, r_offset, r_len, RL_WRITER);
 
+	/*
+	 * TODO: error handling in case of UZFS_IO_TX_ASSIGN_FAIL
+	 */
+	if (zv->in_rebuilding_mode) {
+		if (!is_rebuild)
+			uzfs_add_to_incoming_io_tree(zv, offset, len);
+		if (is_rebuild) {
+			count = uzfs_search_incoming_io_tree(zv, offset,
+			    len, (void **)&chunk_io);
+chunk_io:
+			if (count) {
+				node = list_remove_head(chunk_io);
+				offset = node->offset;
+				len = node->len;
+				end = offset + len;
+				umem_free(node, sizeof (*node));
+				wrote = offset - orig_offset;
+				len_in_first_aligned_block = 0;
```

Comment:
len_in_first_aligned_block is getting set to zero if count is non-zero. I think we need to handle it in cases where we have overlapping writes.

Reply:
yes.. setting this to 0 is not correct.. it should be:
        r_offset1 = (offset / blocksize) * blocksize;
 	len_in_first_aligned_block = (blocksize - (offset - r_offset1));
 
 	if (len_in_first_aligned_block > len)
                len_in_first_aligned_block = len;

Reply:
but, with len_in_first_aligned_block being 0 also, things works fine. This variable makes sure that IOs are aligned with block size even if initial offset is misaligned to blocksize.

Reply:
ok. I will make changes accordingly.